packages/apollo-server-core/src/requestPipeline.ts

```diff
@@ -162,35 +164,64 @@ export async function processGraphQLRequest<TContext>(
     requestContext,
   });
 
-  const parsingDidEnd = await dispatcher.invokeDidStartHook(
-    'parsingDidStart',
-    requestContext,
-  );
-
   try {
-    let document: DocumentNode;
-    try {
-      document = parse(query);
-      parsingDidEnd();
-    } catch (syntaxError) {
-      parsingDidEnd(syntaxError);
-      return sendErrorResponse(syntaxError, SyntaxError);
+    // If we're configured with a document store, we'll utilize the hash of the
+    // operation in order to lookup a previously parsed-and-validated operation
+    // from that.  A failure to retrieve anything from the cache simply means
+    // we're guaranteed to do the parsing and validation ourselves.
+    let document: DocumentNode | undefined;
+
+    if (config.documentStore) {
+      document = await config.documentStore.get(queryHash);
```

Comment:
If the document store returns an error (ie, the await throws), should we really fail the whole query?

(I see that in practice the document store is in-memory so this probably never happens...)

Reply:
This is worth guarding against, even if it's not currently possible.

Done via 1a9d65dd810e4bd47191c800c9b53a92c1ca6f88.